### DISCUSSION TITLE
Implementing JTE in a Spring Boot Multi-Module Application

### DISCUSSION THREAD

**Discussion by astellj** [casid/jte#273] [Q&A]
Hello everyone!  

I have just started working on a Spring Boot multi-module application, and I want to integrate JTE templates into it. I have previously used JTE in regular applications and enjoy using it, however I am running into some challenges implementing it as part of this multi-module project. Below is the relevant part of my build.gradle file:  

``` 
buildscript {
    ext {
        springBootVersion = '3.1.3'
    }
    repositories {
        mavenCentral()
        gradlePluginPortal()
    }
    dependencies {
        classpath("org.springframework.boot:spring-boot-gradle-plugin:${springBootVersion}")
        classpath("gg.jte:jte-gradle-plugin:3.1.0")
    }

}

subprojects {
    apply plugin: 'java-library'
    apply plugin: 'io.spring.dependency-management'
    apply plugin: 'org.springframework.boot'
    apply plugin: 'gg.jte.gradle'

    ... other config...
}

project(':app') {

    bootRun {
        environment SPRING_PROFILES_ACTIVE: environment.SPRING_PROFILES_ACTIVE ?: 'local'
        workingDir = project(':').projectDir

    }

    dependencies {
        api project(':jetset')
    }
}

project(':jetset') {
    dependencies {
        api('org.springframework.boot:spring-boot-starter-web')
        api('org.springframework.boot:spring-boot-starter-data-jpa')
        runtimeOnly('org.postgresql:postgresql')
        api('org.liquibase:liquibase-core')
        implementation('gg.jte:jte-spring-boot-starter-3:3.1.0')
        implementation('gg.jte:jte:3.1.0')
    }
}

```

This is my basic controller:

```
@SpringBootApplication
@Controller
public class JetsetApplication {

  public static void main(String[] args) {
    SpringApplication.run(JetsetApplication.class, args);
  }

  @GetMapping("/")
  public String index() {
    return "home/index";
  }

}

```

As you can see I have included the JTE & Spring Boot starter dependancy in my ‘jetset’ module, however when I create and attempt to render a template I am presented with a persistant 404 error. I have tried all combinations to attempt to debug, such as creating the templates in the app directly, moving the dependancies around into the app, but have still not found a way to render the templates. I have a controller in both the module and app but neither of them are able to render a template.

Any help at all, or a pointer to it being done somewhere would be much appreciated. Thank you in advance!!

**Comment by robertellam** [comment]
I think I have got this working by defining a JteConfiguration class which uses the ResourceCodeResolver instead of the DirectoryCodeResolver:

```
@Configuration
@EnableConfigurationProperties(JteProperties.class)
public class JteConfiguration {
    private final JteProperties jteProperties;

    public JteConfiguration(JteProperties jteProperties) {
        this.jteProperties = jteProperties;
    }

    @Bean
    @ConditionalOnMissingBean(JteViewResolver.class)
    public JteViewResolver jteViewResolver(TemplateEngine templateEngine) {

        return new JteViewResolver(templateEngine, jteProperties.getTemplateSuffix());
    }


    @Bean
    @ConditionalOnMissingBean(TemplateEngine.class)
    public TemplateEngine jteTemplateEngine() {
        if(jteProperties.isDevelopmentMode() && jteProperties.usePreCompiledTemplates()){
            throw new JteConfigurationException("You can't use development mode and precompiledTemplates together");
        }
        if (jteProperties.usePreCompiledTemplates()) {
            // Templates will need to be compiled by the maven/gradle build task
            return TemplateEngine.createPrecompiled(ContentType.Html);
        }
        if(jteProperties.isDevelopmentMode()) {
            // Here, a JTE file watcher will recompile the JTE templates upon file save (the web browser will auto-refresh)
            // If using IntelliJ, use Ctrl-F9 to trigger an auto-refresh when editing non-JTE files.

            CodeResolver codeResolver = new ResourceCodeResolver(jteProperties.getTemplateLocation());
            return TemplateEngine.create(codeResolver, Paths.get("jte-classes"), ContentType.Html, getClass().getClassLoader());
        }
        throw new JteConfigurationException("You need to either set gg.jte.usePrecompiledTemplates or gg.jte.developmentMode to true ");
    }
}
```

In my application yml i think added:
```gg:
  jte:
    developmentMode: true
    templateLocation: jte
```
    
  I then created two controllers one in the app and one in the jetset module. I then created jte templates in `resources/jte` directories under the app and the module and this now seems to work.
  
  Can anyone confirm is this is the correct approach?

**Comment by DENEB1593** [comment]
I have sample problem on my project with gradle-multi-module

**Comment by ashmelev** [comment]
I had a similar problem which was resolved by changing the `template-location` property.
More specifically, I tend to lay out my multi-module applications such that all code modules are in sub-projects, for example:

```
project-root
  |
  -- api
        |
         -- api-module-1
         -- api-module-2
         -- ui-module-1
               |
               -- src/main/jte/index.jte
  -- lib
        |
         -- custom-lib-1
         .....
```

Given the structure above, the `ui-module-1` is providing the JTE user interface pages and the `template-location` property must be set like: 

```
gg:
  jte:
    development-mode: true
    use-precompiled-templates: false
    template-location: api/ui-module-1/src/main/jte    # MUST be a path rooted at the master project level

```
in order the JTE to find the templates.  Although, in hindsight, this was obvious, it took a good bit of tracing in the debugger to understand what was causing the 404/TemplateNotFound exceptions.  Changing the `template-location` resolved it.

As an IMHO note, I should say this may only be true in development/local usage.  The root of how templates are resolved may change once the `ui-module-1` component is containerized and deployed to Kubernetes (for example) in production.  I've not gotten that far in my development/testing.  That said, we can resolve the issue with dev/qa/prod-aware property files (or similar means), but I'd wonder if there was not a fix/change that could be made to the template resolution code to stabilize the location

**Comment by kicktipp** [comment]
Same problem here. This is annoying and difficult to debug. I had to go all the way, too before I found this ticket. (JTE is great so far otherwise)
